File: Data Structures & Algorithms/subsets/submission-1.py
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        
        result = list()
        def helper(path, curr, length):

        
            if curr >= length:
                result.append(path)
                return

            helper(path + [nums[curr]], curr + 1, length)
            helper(path, curr + 1, length)


        helper(list(), 0, len(nums))

        return result

File: Data Structures & Algorithms/maximum-product-subarray/submission-1.cpp
class Solution {
public:
    int maxProduct(vector<int>& nums) {
        
        std::vector<int> mins(nums.size(), 0);
        std::vector<int> maxs(nums.size(), 0);

        maxs[0] = nums[0];
        mins[0] = nums[0];

        for (size_t i = 1; i < nums.size(); ++i) {
            int num = nums[i];
    
            maxs[i] = std::max({num, maxs[i-1] * num, mins[i-1] * num});
            mins[i] = std::min({num, mins[i-1] * num, maxs[i-1] * num});
    
        }

        int result = INT_MIN;
        for (auto num : maxs){  
            std::cout << num << " ";
            result = std::max(result, num);
        }

        return result;


    }
};
